### PR TITLE
GLOB-20260 Remove winston dependency.

### PR DESCRIPTION
**Pull request by cdm-ium**
Why? Winston includes a number of dependencies for features we don't
use. This needlessly increases the surface area of projects that use it.

This PR is a wip to explore how we could remove the winston dependency.
It still needs more tests and more experimentation in an actual project.

I based my approach to moving winston off of how the very simple morgan
(https://github.com/expressjs/morgan) works. Basically morgan is just a
very lightweight wrapper on the writable stream that is passed to its
constructor. I'm trying the same thing for our logging with console and
loggly streams.

Current state of things:
- Code is mostly working I think this is just about ready for merge pending a little more testing
- basically removed usage of streams because the objects we are using to write logs to outputs are either already streams (e.g. process.stdout) or support buffering (the loggly client in from `node-loggly-bulk`)

For the life of me I couldn't actually get implementing writable to work properly in node. Including the exact example from loggly here: https://www.loggly.com/blog/node-js-libraries-make-sophisticated-logging-simpler/

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -20,8 +23,6 @@
         "lodash": "^4.17.4",
         "morgan": "^1.7.0",
         "morgan-json": "^1.1.0",
-        "winston": "^2.3.0",
```
Comment (cdm-ium) [inline package.json]: Good bye.

File: src/logger.js [hunk 2]
```diff
@@ -78,39 +37,44 @@ function sortObj(obj) {
...
         const { name } = container.metadata;
-        this.baseLogger = createLogger(name, this.config.level, this.config.loggly);
+        this.stream = createLoggerStream(name, this.config.level, this.config.loggly);
         this.requestRules = this.config.requestRules;
     }
 
     debug(req, message, args, autoLog = true) {
         const params = this.createLogParameters(req, message, args, autoLog);
-        this.baseLogger.debug(message, params);
+        params.level = 'debug';
```
Comment (saivickna) [inline src/logger.js]: add `level` as a parameter in createLogParameters?

File: package.json [hunk 3]
```diff
@@ -4,6 +4,9 @@
     "description": "Opinionated logging for Node projects",
     "main": "lib",
     "repository": "https://github.com/globality-corp/nodule-logging",
+    "jest": {
```
Comment (cdm-ium) [inline package.json]: I remember seeing this error in other projects when we updated some library versions. Not sure why it appeared now expect maybe some libraries were way out of date.

File: src/streams/LogglyStream.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import { Writable } from 'stream';
+
+import loggly from 'node-loggly-bulk';
```
Comment (saivickna) [inline src/streams/LogglyStream.js]: How come this is not in `package.json`?

Comment (cdm-ium) [reply src/streams/LogglyStream.js]: Just forgot to add it.

File: src/middleware.js [hunk 5]
```diff
@@ -5,17 +5,6 @@ import { get, set } from 'lodash';
 import omitBy from 'lodash/omitBy';
 
 
-// where morgan connects to winston
-function addStream(logger, level) {
```
Comment (cdm-ium) [inline src/middleware.js]: I never understood the meaning of this function before but after reading the morgan code I do. morgan isn't designed to work with a full-fledged logger, but a stream (basically an object with an interface that exposes a `write` function). We were passing morgan a logger which we were wrapping as a stream (exposing a `write` function) on a logger ojbect that itself probably had a stream somewhere under the hood. We were adding a few layers of direction and it probably wasn't necessary. Simply removing it will remove `level` from the logs morgan generates so I will go back and fix that.

File: src/__tests__/logger.test.js [hunk 6]
```diff
@@ -14,35 +11,9 @@ describe('create a new logger and transports', () => {
...
-    });
-
-
     it('should not fail if graph is not initialized', () => {
         clearBinding('logger');
         const logger = getLogger();
         logger.info({}, 'hello there');
     });
-
-    it.skip('should return a writable stream for morgan to write logs to', () => {
```
Comment (cdm-ium) [inline src/__tests__/logger.test.js]: We don't need a skipped test for a function that was removed.

File: src/streams/LogglyStream.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            json: true,
+        });
+
+        // make sure write function has proper context (this)
+        this._write = this._write.bind(this); // eslint-disable-line
+    }
+
+    _write(message, encoding, callback) { // eslint-disable-line
+        this.client.log(JSON.parse(message));
+        /* XXX callback causing server exit??
```
Comment (cdm-ium) [inline src/streams/LogglyStream.js]: I don't know what this callback does, but calling it seems to cause the server to exit. I did check the data the callback is being called with and we are getting a successful response from loggly...

File: src/streams/UnionStream.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+        this.streams = options.streams;
+
+        // make sure write function has proper context (this)
+        this._write = this._write.bind(this); // eslint-disable-line
+    }
+
+    _write(message, encoding, callback) {
+        if (typeof message === 'object') {
+            this.streams.map(stream => stream.write(JSON.stringify(message), encoding, callback));
```
Comment (cdm-ium) [inline src/streams/UnionStream.js]: I have to unravel the `json.stringify/parse` chaining that I'm doing here. I think what I'm doing is overly complicated.

File: src/middleware.js [hunk 9]
```diff
@@ -24,6 +13,16 @@ function skip(ignoreRouteUrls) {
     };
 }
 
+function asStream(logger) {
```
Comment (cdm-ium) [inline src/middleware.js]: I readded the stream function removed above. It's the most convenient way to make sure that the logs morgan generates receive a log-level.

The morgan documentation says it wants a stream passed to the middleware but in the original version of this code (with `addStream`) we didn't actually pass a stream and instead just wrapped the logger in an object with a `write` property because that is the api morgan actually requires. It worked for us in the past so I assume it won't cause problems.

File: src/streams/UnionStream.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
+import { Writable } from 'stream';
+
+class UnionStream extends Writable {
```
Comment (cdm-ium) [inline src/streams/UnionStream.js]: I'm going to try rewriting this as not a stream tomorrow because after playing around with extending streams for the `loggly stream` I don't really see the benefit of having a stream here because the union stream (as used) is a union of `process.stdout` which is a stream and has all the functions stream has and the loggly client which I think does it's own buffering internally anyways. I will clean up naming tomorrow.

File: src/streams/UnionStream.js (hunk 8 above)
Comment (cdm-ium) [reply src/streams/UnionStream.js]: Before killing the stream implemntation altogether I want to change the `stream.write` calls here to `stream._write` which might fix my issues with the callback passing.
